app/src/main/java/tech/relaycorp/gateway/data/disk/DiskOperations.kt

```diff
@@ -0,0 +1,74 @@
+package tech.relaycorp.gateway.data.disk
+
+import android.content.Context
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import java.io.File
+import java.io.IOException
+import java.io.InputStream
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
+class DiskOperations
+@Inject constructor(
+    private val context: Context
+) {
+
+    @Throws(DiskException::class)
+    suspend fun writeMessage(folder: String, prefix: String, message: ByteArray): String =
+        try {
+            writeMessageUnhandled(folder, prefix, message)
+        } catch (e: IOException) {
+            throw DiskException(e)
+        }
+
+    suspend fun listMessages(folder: String): List<() -> InputStream> =
+        withContext(Dispatchers.IO) {
+            getOrCreateDir(folder)
+                .listFiles()
+                ?.map { it::inputStream }
+                ?: emptyList()
+        }
+
+    @Throws(ParcelDataNotFoundException::class)
+    suspend fun readMessage(folder: String, path: String): (() -> InputStream) {
+        val file = File(getOrCreateDir(folder), path)
+        if (!file.exists()) throw ParcelDataNotFoundException("Parcel data not found on path '$path'")
```

Comment:
I'd suggest to be consistent: Is this class/method dealing with messages in general or parcels specifically?

Reply:
Oh right, that's a mistake, it's a generic helper class, so messages.